lib/kube/grpc/grpc.go

```diff
@@ -0,0 +1,314 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpc
+
+import (
+	"context"
+	"errors"
+
+	"github.com/gravitational/trace"
+	"github.com/gravitational/trace/trail"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/gravitational/teleport/api/defaults"
+	proto "github.com/gravitational/teleport/api/gen/proto/go/teleport/kube/v1"
+	"github.com/gravitational/teleport/api/types"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// errDone indicates that resource iteration is complete
+var errDone = errors.New("done iterating")
+
+// Server implements KubeService gRPC server.
+type Server struct {
+	proto.UnimplementedKubeServiceServer
+	cfg          Config
+	proxyAddress string
+	kubeProxySNI string
+}
+
+// New creates a new instance of Kube gRPC handler.
+func New(cfg Config) (*Server, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	sni, addr, err := getWebAddrAndKubeSNI(cfg.KubeProxyAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &Server{cfg: cfg, proxyAddress: addr, kubeProxySNI: sni}, nil
+}
+
+// Config specifies configuration for Kube gRPC server.
+type Config struct {
+	// Context passes the optional external context
+	// passing global close to all operations.
+	Context context.Context
+	// AuthClient is a auth server client.
+	AuthClient auth.ClientI
+	// Authz authenticates user.
+	Authz auth.Authorizer
+	// Log is the logger function.
+	Log logrus.FieldLogger
+	// Emitter is used to emit audit events.
+	Emitter apievents.Emitter
+	// Component name to include in log output.
+	Component string
+	// KubeProxyAddr is the address that can be used to reach the Kubernetes Proxy.
+	KubeProxyAddr string
+	// ClusterName is the name of the cluster that this server is running in.
+	ClusterName string
+}
+
+// CheckAndSetDefaults checks and sets default values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.AuthClient == nil {
+		return trace.BadParameter("missing parameter AuthClient")
+	}
+	if c.Authz == nil {
+		return trace.BadParameter("missing parameter Authz")
+	}
+	if c.Emitter == nil {
+		return trace.BadParameter("missing parameter Emitter")
+	}
+	if c.Context == nil {
+		c.Context = context.TODO()
```

Comment:
Nit: part of me wonders if it's nicer to just error if no parent context is provided ? or use `context.Background` here ? I tend to reserve `context.TODO()` for situations where future refactoring could be completed to add `context.Context` to the parameters of a function.